rust/src/buffer.rs

```diff
@@ -20,18 +20,25 @@ use libc;
 use std::mem;
 use std::slice;
 
+use super::datatypes::*;
 use super::memory::*;
 
 /// Buffer<T> is essentially just a Vec<T> for fixed-width primitive types and the start of the
 /// memory region is aligned at a 64-byte boundary
-pub struct Buffer<T> {
+pub struct Buffer<T>
```

Comment:
I'm curious about the design choice here. In the C++ world, Arrow buffers are typeless (they just point to spans of bytes), it's arrays that give them their meaning. Do you benefit from giving buffers a specific type?

Reply:
I think I do. I have type-safe Builder<T> where the push() method does not need to do any conversion per value being pushed, so it seems as efficient as possible. Then the memory is passed to the Buffer<T> where looking up values by index also seems as efficient as possible without any calculations on each lookup.

I am spending more time looking at the C++ implementation now to try and learn from it and keep the Rust version consistent. If you think there are arguments for using a Buffer<u8> instead of Buffer<T> I'd like to hear those too.

Reply:
It seems this is based on the assumption that other languages have no problem interpreting the memory layout used to represent `T` in Rust, right? For instance, `&'static str` is a `ArrowPrimitiveType` - what's its memory layout look like? can it be correctly interpreted in languages such as Java?

Another thing is, this makes it hard to use a different implementation for a particular `T`. For instance, if we want to [use bit-packing for boolean type](https://issues.apache.org/jira/browse/ARROW-2483), then this doesn't allow it, since `bool` always occupy 1 byte.

IMHO `Buffer` should only care about how to represent a continuous memory segment. Interpreting the memory layout into different logical types should be handled in a different module (e.g., `array`).

Reply:
So the overall question here is: does `Buffer` know the memory layout for primitive types or does it not? And what are primitive types? For example --- and as @sunchao pointed out correctly --- strings are not that primitive, since the memory layout differs heavily amongst languages (e.g. C uses a pointer to a char-array where the last character is `'\0'`, C++ stores the pointer to a char-array and it's length but in implementation-definied order, some Java versions use singelton objects and so on). And even if the types are primitive and are not packed (like booleans for example), we still have the [endiannes](https://en.wikipedia.org/wiki/Endianness) problem (see  [Specs](https://arrow.apache.org/docs/memory_layout.html#byte-order-endianness)). So I think we may need to implement the conversion and read/write routines for every type and may be able to use macros for common cases (like `u8`, `u16`, `u32`, `u64` are very similar).

Since we now have memory-IO for every type, the `Buffer` could be typeless. I think the conversion-free buffer-to-buffer copies are something that sound great, but I think the `Array`-type should deal with it, not the buffer. That keeps the design somehow clean and we don't need to repeat the types in that many layers.

Reply:
OK so if I understand we want to use Buffer<u8> for all memory rather than `Buffer<T>` and then have Array take care of the type conversions. That makes sense to me, but again I think that is a separate PR from converting array from enum to trait/generic. I'm concerned that we're expanding the scope of this particular PR.